RecoLocalCalo/HcalRecAlgos/src/MahiFit.cc

```diff
@@ -0,0 +1,633 @@
+			  float& reconstructedTime,
+			  bool& useTriple, 
+			  float& chi2) const {
+
+  assert(channelData.nSamples()==8||channelData.nSamples()==10);
+
+  resetWorkspace();
+
+  nnlsWork_.tsSize = channelData.nSamples();
+  nnlsWork_.tsOffset = channelData.soi();
+  nnlsWork_.fullTSOffset = fullTSofInterest_ - nnlsWork_.tsOffset;
+
+  // 1 sigma time constraint
+  if (channelData.hasTimeInfo()) nnlsWork_.dt=timeSigmaSiPM_;
+  else nnlsWork_.dt=timeSigmaHPD_;
+
+  //Dark current value for this channel (SiPM only)
+  float darkCurrent =  getSiPMDarkCurrent(channelData.darkCurrent(), 
+					  channelData.fcByPE(),
+					  channelData.lambda());
+
+  //Average pedestal width (for covariance matrix constraint)
+  float pedVal = 0.25*( channelData.tsPedestalWidth(0)*channelData.tsPedestalWidth(0)+
+			channelData.tsPedestalWidth(1)*channelData.tsPedestalWidth(1)+
+			channelData.tsPedestalWidth(2)*channelData.tsPedestalWidth(2)+
+			channelData.tsPedestalWidth(3)*channelData.tsPedestalWidth(3) );
+
+  nnlsWork_.pedConstraint = pedVal*SampleMatrix::Ones(nnlsWork_.tsSize, nnlsWork_.tsSize);
+  nnlsWork_.amplitudes.resize(nnlsWork_.tsSize);
+  nnlsWork_.noiseTerms.resize(nnlsWork_.tsSize);
+
+  std::vector<float> reconstructedVals;
+  
+  double tsTOT = 0, tstrig = 0; // in GeV
+  for(unsigned int iTS=0; iTS<nnlsWork_.tsSize; ++iTS){
+    double charge = channelData.tsRawCharge(iTS);
+    double ped = channelData.tsPedestal(iTS);
+
+    nnlsWork_.amplitudes.coeffRef(iTS) = charge - ped;
+
+    //ADC granularity
+    double noiseADC = (1./sqrt(12))*channelData.tsDFcPerADC(iTS);
+
+    //Dark current (for SiPMs)
+    double noiseDC=0;
+    if(channelData.hasTimeInfo() && !channelData.hasEffectivePedestals() && (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noiseDC = darkCurrent;
+    }
+
+    //Photostatistics
+    double noisePhoto = 0;
+    if ( (charge-ped)>channelData.tsPedestalWidth(iTS)) {
+      noisePhoto = sqrt((charge-ped)*channelData.fcByPE());
+    }
+
+    //Electronic pedestal
+    double pedWidth = channelData.tsPedestalWidth(iTS);
+
+    //Total uncertainty from all sources
+    nnlsWork_.noiseTerms.coeffRef(iTS) = noiseADC*noiseADC + noiseDC*noiseDC + noisePhoto*noisePhoto + pedWidth*pedWidth;
+
+    tsTOT += charge - ped;
+    if( iTS==nnlsWork_.tsOffset ){
+      tstrig += (charge - ped)*channelData.tsGain(0);
+    }
+  }
+
+  if(tstrig >= ts4Thresh_ && tsTOT > 0) {
+
+    useTriple=false;
+
+    // only do pre-fit with 1 pulse if chiSq threshold is positive
+    if (chiSqSwitch_>0) {
+      doFit(reconstructedVals,1);
+      if (reconstructedVals[1]>chiSqSwitch_) {
+	doFit(reconstructedVals,0); //nbx=0 means use configured BXs
+	useTriple=true;
+      }
+    }
+    else {
+      doFit(reconstructedVals,0);
+      useTriple=true;
+    }
+  }
+  else{
+    reconstructedVals.clear();
+    reconstructedVals.push_back(0.);
+    reconstructedVals.push_back(888.);
+    reconstructedVals.push_back(888.);
```

Comment:
please use the same default values as in M2, unless there is a good reason to be different
https://github.com/cms-sw/cmssw/blob/master/RecoLocalCalo/HcalRecAlgos/src/PulseShapeFitOOTPileupCorrection.cc#L317-L321
the choice there is -9999 instead of 888.